client/pdf_generator.py

```diff
@@ -0,0 +1,175 @@
+#!/usr/bin/python3
+import time
+import os
+
+from reportlab.lib.enums import TA_JUSTIFY
+from reportlab.lib.pagesizes import letter
+from reportlab.platypus import SimpleDocTemplate, Paragraph, Spacer, Image
+from reportlab.lib.styles import getSampleStyleSheet
+from reportlab.lib.units import inch
+from reportlab import platypus
+from  reportlab.lib.styles import ParagraphStyle as PS
+
+import king_phisher.client.mailer as mailer
+import king_phisher.client.plugins as plugins
+
+import jinja2.exceptions
+
+def _expand_path(outfile, joins, pathmod=os.path):
+	outfile = pathmod.expandvars(outfile)
+	outfile = pathmod.expanduser(outfile)
+	outfile.join(outfile, *joins)
+	return outfile
+
+class Plugin(plugins.ClientPlugin):
+	authors = ['Jeremy Schoeneman']
+	title = 'Generate PDF'
+	description = """
+	Generates a PDF file with a link which includes the campaign url with the individual 
+	message_id required to track individual visits to a website. Visit 
+	https://github.com/y4utj4/pdf_generator for example template files to use for this 
+	plugin
+	"""
+	homepage = 'https://github.com/securestate/king-phisher-plugins'
+	options = [
+		plugins.ClientOptionString(
+			'output_pdf',
+			'pdf being generated',
+			default='$HOME/Attachment1.pdf',
+			display_name='* Output PDF File'
+		),
+		plugins.ClientOptionString(
+			'template_file',
+			'Template file to read from',
+			default='$HOME/template.txt',
+			display_name='* Template File'
+		),
+		plugins.ClientOptionString(
+			'logo',
+			'Image to include into the pdf',
+			display_name="Logo Image"
+		),
+		plugins.ClientOptionString(
+			'link_text',
+			'Text for inserted link',
+			default='Click here to accept',
+			display_name="Link Text"
+		)
+	]
+	req_min_version = '1.7.0b1'
+	version = '1.1'
+
+	def initialize(self):
+		mailer_tab = self.application.main_tabs['mailer']
+		self.text_insert = mailer_tab.tabs['send_messages'].text_insert
+		self.add_menu_item('Tools > Create PDF Preview', self.make_preview)
+		self.signal_connect('send-precheck', self.signal_send_precheck, gobject=mailer_tab)
+		self.signal_connect('send-target', self.signal_send_target, gobject=mailer_tab)
+		self.signal_connect('send-finished', self.signal_send_finished, gobject=mailer_tab)
+		return True
+
+	def signal_send_precheck(self, _):
+		if not any((self.config['template_file'], self.config['output_pdf'], self.config['link_text'])):
+			self.logger.debug('skipping exporting any attachments due to lack of information provided to run')
+			return True
+
+		if not os.path.isfile(self.config['template_file']):
+			self.logger.error('no pdf template file found')
+			return
+		else:
+			self.logger.debug('pdf template file found, generating attachment')
+			return True
+
+		if self.config['logo']:
```

Comment:
This and the code below it will never get executed because the conditional above has `return` statements in both branches. This should probably also be `if self.config['logo'] and not os.path.isfile(...`.